### PR TITLE
Added the Automatic Tracking of Award Eligibility and Additional Award Enhancements

### LINKED ISSUE #851: Assign Award to all personnel in Force
Just as a neat idea, if you could assign awards to all the persons in a force in the TOE panel, that would be really cool. That way you could quickly assign awards to all members of each lance or company that earned them without needing to hunt around the personnel panel.

This would also pave the way for unit level commendations which could piggy back most of the same code.

### LINKED ISSUE #871: [0.45.0 : Feature Request] Enhance after combat resolution to support medal/ribbon awards
Enhance after combat resolution to support medal/ribbon awards. 

1. Show total kills per individual/unit, hits taken or any other ‘special’ achievement that can be captured from MM round report

2. Auto suggest medals/ribbons

3. A new tab in campaign setup to setup medals/ribbons and its requirements

This will make life a lot easier for big games.

Thanks.

### LINKED ISSUE #2429: Error loading campaign data in AwardTracker
### Environment
What version of MekHQ does your issue apply to? 0.47.17
What operating system are you using?  Windows
What java version are you using?  	Java vendor AdoptOpenJDK 11.0.9.1

### Description
While updating the AwardTracker with my campaign info, I get an error saying the 'date-type' column has duplicate entries.  None of the dates/scenario names match. 

### Files
Attached zip contains my campaign file and award tracker file
[AwardTracker.zip](https://github.com/MegaMek/mekhq/files/6006307/AwardTracker.zip)
[customs.zip](https://github.com/MegaMek/mekhq/files/6006309/customs.zip)

### LINKED ISSUE #2433: RFE: Save kill summary on mission completion
### Description
When trying to determine awards, it is relatively straightforward to manually check the pilots involved in a single combat and see whether or not they should be granted awards for their performance.

However, when attempting to determine _mission_-scale awards for individuals, lances, etc it is much more complicated as you need to check dates and count kills over _many_ separate combats as well as then perform math over multiple scales (lances, companies, etc).

Would it be possible to have the "Complete Mission" also save a summary of the combat that occurred during the mission you are completing?

Having the data in the same format as the TO&E layout would be great:
- Main Force : Mission : total kills
- - Sub Force 1 : Mission : total kills
- - - Pilot/Unit 1 : Mission : kills
- - - Pilot/Unit 2 : Mission : kills
- - Sub Force 2 : Mission : total kills
- - - Pilot/Unit 3 : Mission : kills
- - - Pilot/Unit 4 : Mission : kills
etc

### LINKED ISSUE #2772: RFE: Menu tree structure for awards
I'm currently created 858 ribbons because I don't want to use real military badges (like the actual ones) in my game. Unfortunately, this sheer number cannot be used properly in the existing structure.

I would therefore like to see an enhancement in the following points so that I can implement the ribbons for everyone. I've already added the graphics of the ribbons. 

**Point 1:**
I would like the tree menu of Awards to have more ramifications so that the structure does not look like the attached picture, but have the possibility of an intermediate branch. 

So instead of the current one "Academy-CC> Capella War College I", a structure like "Academy> CC> Capella War College I" would be great.

Award Menu

[Awards.zip](https://github.com/MegaMek/mekhq/files/6891728/Awards.zip)

### LINKED ISSUE #2773: RFE: Display options for awards
I'm currently created 858 ribbons because I don't want to use real military badges (like the actual ones) in my game. Unfortunately, this sheer number cannot be used properly in the existing structure.

I would therefore like to see an enhancement in the following point so that I can implement the ribbons for everyone. I've already added the graphics of the ribbons. 

**Point 2:**
I have created visual changes to the ribbons for a large number of awards, if the award was awarded multiple times, for example as with the Silver Star (see attached graphic).

Silver Star

Therefore, if there were awards that could be awarded multiple times, I would like to have the option of defining which image and text should be used for the actual number of the award. 

After all, only 28 ribbons can be displayed under the portrait and it makes no sense if these were nine different Silver Star Ribbons, among other things. 

Thank you!

### LINKED ISSUE #3495: Quality of life request: please re-order the default awards list
The default awards list is difficult to navigate, as less frequently applied awards are at the top (e.g. Medal of Honor) while more commonly awarded ones are buried somewhere in the middle (e.g. Purple Heart).  Additionally kill-count based awards tend to be in reverse order e.g. higher kill counts before lower ones.  There is an order element in the XML file however it is not being made use of.

It would be helpful if the order was changed to reflect how often they are assigned.  For example, like this:

Altered awards order

(txt extension added to allow attachment)
[standard.xml.txt](https://github.com/MegaMek/mekhq/files/9955059/standard.xml.txt)

### LINKED ISSUE #3623: RFE: improve handling of wide medals
According to the Award Guide.md help file:

`Medal images must be in a transparent .png format with an 87-pixel height.`

Medal images are being scaled based on width, rather than height, with images that exceed the maximum width being scaled down.  This means that items that are wide rather than tall, such as jump wings, are difficult to add - making them wide enough to be legible only gets them shrunk.

It would be helpful if the medal images were scaled to height with the width allowed to float, in line with the help file description.  Not sure if the image layout can handle that automatically or if there needs to be additional code to space them out properly.

### LINKED ISSUE #4024: RFE Awards Suggestions
The awards change currently in process is some thing I have been waiting on for ever since awards were added. This RFE is also related to awards. 

Currently there are awards within the standard set which can be awarded to an individual multiple times. However nothing changes visually with the award ribbon or medal. The US military awards has devices which are added to the ribbon / medal for each subsequent award of the same medal. Depending on the medal this is either a gold star or bronze star, with 5 gold stars being replaced by a silver star. Numerals are also used for some awards. There are a pretty large number of different devices which can be used.  

This RFE asks that these devices be visually represented for subsequent awards of the same award. I've added an image as an example of some of the devices which could be used. 

awards

### PR DESCRIPTION
**Pull request by IllianiCBT**
## Introduction
autoAwards adds to mhq a new way to track Award eligibility. It does this by separating Awards into types and by utilizing the previously unused fields `qty`, `item`, `size`, and `range`. It also introduces a couple of functions to help read the TOE, something that is essential to autoAwards, but will likely be useful in other parts of the suite.

This PR covers a _lot_, but I will do my best to cover everything.

**Note:** this PR is a WIP, so some functionality may change before launch. As additions or edits are made, these will be reflected in the text.

**Terminology:** Mission is synonymous with Contract and AtBContract, any described functionality applies to all three. Contracts are synonymous with AtBContracts, any described functionality applies to both.. If AtBContract is specified, the functionality only applies to AtBContracts. `autoAwards` is the internal name for the system that automatically tracks Awards eligibility and is referred to as such throughout this PR.

**Testing:** this is the largest & most complicated coding project I have done to date. While I am proud of what I've accomplished, it is important to recognize my limited experience. The functionality implemented by autoAwards should be _thoroughly_ tested before launch.

## Trigger
autoAwards is triggered by the conclusion of Missions, Scenarios, and manually through the toolbar.

## Documentation
A documentation pdf has been provided which replaces the previous readme. I have also moved the old Awards documentation & spreadsheet into the archive folder as functionally it is replaced by autoAwards.

## Award Types
autoAwards currently offers support for 12 types of Award, with room for others to be added down the line. In each case the `item` field tells autoAwards what type of Award it is looking at. Awards are grouped according to their type, which ensures that we're not parsing the same Awards multiple times unnecessarily.

As mentioned above, autoAwards uses the previously unused `qty`, `item`, `size`, and `range` fields already included in the Award dataset.

image

### Contract Awards (Contracts only)
These awards are issued at the conclusion of a contract if the contract type matches the type listed in the Award, or if the contract's duration matches `qty`. If the Mission is not a Contract or AtBContract, these Awards are skipped.

`qty` is used only with the duration sub-type. If `range` is set to `months`, then autoAwards checks to see whether the true duration (start date to current date) of the contract (in months) equals or exceeds this value.
`size` is ignored, with no use-case intended at this time.
`range` tells autoAward what type of Contract needs to be completed. This can be either the literal contract type, or `raid` or `duty` which will be true if the Contract was a raid or duty contract, respectively. Alternatively, this value can be set to 'months' to tell autoAwards to check the Contract's true duration, instead.

### Faction Hunter Awards (AtBContracts Only)
These awards are issued for completing contracts against specific enemy factions. If the Mission is not an AtBContract (and AtB is enabled), these Awards are skipped.

`qty` is currently ignored, although I can see me adding support to allow users to specify that their Award is issued only after completing x number of Contracts against y enemy.
`size` is ignored, with no use-case intended at this time.
`range` tells autoAwards what type of enemy needs to be faced. This can either be the Faction name (for single Factions), or a super-faction, such as 'Clans' which will return true if the faction has the `Clan` tag. Support for checking for multiple Factions is included. So users can specify Award eligibility if any one of a list of Factions is faced.

**Limitations:** Garrison Contracts will trigger multiple different type of Faction opponent. autoAwards does not, currently, know how to handle that and will ignore all Factions except the last Faction faced within the Contract.

### Injury Awards (Missions)
Injury Awards are issued at the end of a Scenario when a person has suffered sufficient Hits during that Scenario.

`qty` is how many Hits must be suffered, during the Scenario, to qualify for the Award
`size` is ignored, with no use-case intended at this time.
`range` is ignored, with no use-case intended at this time.

**Limitations:** autoAwards has no way to tell how a Hit was sustained, so it's entirely possible for personnel to qualify for Injury Awards by falling over a few times.

### Kill Awards (Missions)
These Awards are awarded for killing x number of opposing units at the individual+ level.

`qty` tells autoAwards how many units need to be killed
`size` states what organizational tier the kills are counted at. This ranges from `Individual`, for personal kills, all the way up to `Army`. Technically, it wouldn't be difficult to add support for TOEs larger than Army, but I didn't think it was worth considering at this time. Capping at Army should cover 99% of use cases.
`range` tells autoAwards whether to count kills accumulated across lifetime, the last mission, or across a scenario. Scenario kills are handled separately.

This functionality is achieved through limited TOE reading (see later).

### Rank Awards
These Awards are given out for reaching specific Ranks.

`qty` tells autoAwards what rank numeric needs to be reached to be eligible for the Award. Rank Numerics are a pre-existing measurement that ranges from 0-40 and directly correlates to Rank row number. E8, for example, is rank numeric 8; while WO1 is rank numeric 21, and so on.
`size` is ignored, with no use-case intended at this time.
`range` determines whether the award should consider eligibility based on inclusive Rank, in which case the personnel just needs to met or beat the rank numeric. Or, whether it should be based on exclusive Rank, where NCOs are not eligible for WO or Officer Awards; and visa versa.

### Scenario Awards
These Awards are issued for completing x number of scenarios

`qty` tells autoAwards how many scenarios need to be completed.
`size` is ignored, with no use-case intended at this time.
`range` is ignored, with no use-case intended at this time.

### Scenario Kill Awards
These Awards are issued at the end of a Scenario if the personnel has achieved sufficient kills. These are functionally identical to normal Kill Awards.

Due to being issued after a Scenario (as opposed to after a Mission) additional code support needed to be provided, chiefly streaming a lot of the functionality from AutoAwardsController that isn't needed for Scenario Kill Awards.

### Skill Awards
These Awards are given out for reaching specific Skill Levels.

`qty` tells autoAwards the required Skill Level.
`size` is ignored, with no use-case intended at this time.
`range` is the skill, or list of skills, needed for Award eligibility. If multiple Skills are listed all must meet the required Skill Level for the personnel to be eligible for the Award. Super-skills are included. These include groups of Skills where the personnel is only required to met the requirement for one of the listed Skills. An example of this is the 'Tech' super-skill, which tells autoAwards that the person needs to met the required Skill Level for any one of the Tech Skills. Super-Skills and single Skills can be combined, in which case autoAwards will check to see if the personnel meets the requirement for any *one* Skill covered by the Super-Skill *and also* any listed individual Skills.

### Theatre of War Awards (Contracts only, extra functionality for AtBContracts)
These Awards are issued for accepting Contracts from belligerents during a time of war.  If the Mission is not a Contract or AtBContract, these Awards are skipped.

`qty` is ignored, with no use-case intended at this time.
`size` is used to dictate the period of war. For example `<size>3025,3026</size>` would tell autoAwards that the Award is only valid during those years. It is possible to include the same year twice, such as `<size>3025,3025</size>` which tells autoAwards the Award is only valid during that year.
`range` lists the Belligerents. For Awards intended for use with Contracts only one Faction needs to be listed. In which case autoAwards will check this Faction matches the Employer. However, if the Mission was an AtBContract (and AtB is enabled) additional Belligerents can be included so long as they are followed with `(1)` or `(2)` to dictate which side each Faction falls on. So long as the employer Faction matches one of the Factions on side 1 or 2; and the enemy Faction matches a Faction on the opposing side, the employee may be eligible for this Award. Like _Faction Hunter Awards_ super-factions are supported. Awards with multiple Factions are skipped unless the Mission is an AtBContract and AtB is enabled.

As this is a pretty complicated Award Type, I've included an example below

<p align=center>
image
</p>

### Time Awards
These Awards are issued for serving in a unit for a specific duration. Such as 'serve x years without disciplinary action'.

The existing value `stackable` tells autoAwards whether the required duration is cumulative. If `true` the required duration is increased by the number of times this Award has been issued. Otherwise, `qty` only needs to be beaten once.

`qty` is how many years need to be served
`size` is ignored, with no use-case intended at this time.
`range` is ignored, with no use-case intended at this time.

### Training Awards
_Not currently implemented_

### Misc Awards
These are hard coded Awards that achieve specific goals. These Awards have to be programmed individually, so their implementation is limited.

Currently, Misc Awards cover 6 different types of Award, as described in the documentation.

### Ignored Awards
By setting `item` to `ignore` autoAwards will ignore that specific Award. This is useful in cases where users want to disable a single Award, instead of an entire Award Type. It's also useful for those users that create empty Awards to act as separators between groups of awards.

## Campaign Options
A number of campaign options have been implemented to help users tweak their autoAwards experience.
<p align=center>
image
</p>

`Award Bonuses` allows users to remove the Edge & XP bonuses given from certain Awards (those set up in the .xml to give bonuses). This can be set to one of four options: _Both_ Awards can give both XP and Edge bonuses; _XP_ Awards can only give XP bonuses, any Edge bonuses in the .xml are ignored; _Edge_ Awards can only give Edge bonuses, any XP bonuses in the .xml are ignored; _None_ Awards do not give bonuses, any XP or Edge bonuses in the .xml are ignored
`Award Tier Size` tells mhq how many times an award needs to be issues to a single person, before Person View will start displaying the next available image for that award.
`Track award eligibility` turns autoAwards on & off. Below this are individual checkboxes for each type of Award currently implemented. This allows users to disable entire groups of Awards, should they wish. This option, and all sub-options are enabled by default. If `Track award eligibility` is disabled all sub-options (and the three remaining options, described below) are all disabled.
`Issue posthumous awards` toggles whether autoAwards should consider dead people eligible for Awards. Disabled by default
`Only issue the best award` tells autoAwards whether it should ignore Award eligibility for an Award, if the person is also eligible for a better Award of the same type. This functionality is enabled for Contract Awards (duration sub-type only), Kill Awards (Scenario and General), Injury Awards, Scenario Awards, and Time Awards. For Non-Scenario Kill Awards each formation tier is considered to be a different Award Type, for the purposes of finding the 'best' Award. So multiple Lance Awards are compared against each other, but Lance and Company Awards are not.
`Ignore Standard Set` tells autoAwards to ignore the default Award Set provided by mhq. This is to discourage people from overwriting the Standard Set with their own Custom Sets, potentially creating a troubleshooting nightmare.

## GUI Support
autoAwards has full GUI support, presenting users with a dialog that allows them to dictate which Awards are issued to whom. All columns are sortable, and there is a personnel filter provided that uses the same code as the filter used on the personnel mhq tab. This ensures any code updates will be reflected in autoAwards.

<p align=center>
image
</p>

As seen above, the dialog is currently using AI generated placeholder art. Due to the large amounts of data displayed, autoAwards will automatically expand to full screen size. This removes any annoying resizing that would need to be done on a page-by-page basis.

Speaking of pages, each Award Type is presented separately, to avoid having too much information on display at any one time. The GUI features the ability to `Skip` the current Award Type (without issuing Awards) or to `Skip All`, which skips the current Award Type and all remaining Award Types. `Issue Awards` does what it says on the box.

## Logging
Currently autoAwards is _very_ verbose, over-posting to the mhq log. This is a deliberate design choice intended for the testing period only. Once initial testing has concluded, the majority of log entries will be moved to debug.

## Load
autoAwards is very lightweight and has been stress tested with multiple battalions of units, with supporting staff and hangers on. Testing was performed both with real and constructed data, on a very old laptop. autoAwards completed all required tasks in less than a second.

Actual numbers for the real data were 51 Awards across a total of 458 personnel, for a total of 23,358 Awards processed.

## Reading the TOE
I have added two new functions to `Force.java` that allow for limited reading of a user's TOE.

### getMaximumDepth()
This function uses a recursive search to find the depth of the force furthest from the origin force. It then returns that value (depth) as an integer.

### getDepth()
This function finds the distance (depth) of the Force from the origin Force.

### How is this useful?
By finding the depth of the Force furthest from the Origin Force we can make a couple of assumptions. First, we can assume that any Force at that depth is a Lance (or regional equivalent). If that depth == Lances, then the level above must be a Company (or regional equivalent) and so on, all the way up to (and beyond) Army.

If integrated into other systems, this will allow for improved TOE reading for things like StratCon or AtB scenario generation. Potentially bringing an end to the issue of adding units to company+ forces results in all child forces (and units) being considered for BV balancing.

## Awards from the Standard Set
As I've worked on this project I have been updating the Awards from the Standard (default) set to ensure they're supported by autoAwards.

Furthermore, I went ahead and reorganized the Standard Set so that Awards that have to be manually issued are at the top. I also added fancy group Awards that allow mhq to use sub-menus.

<p align=center>
[image
</p>

Finally, I removed any unnecessary tags that are no longer being used.

These changes will have no negative impact on already issued Awards in pre-existing campaigns.

## Award Images
Previously, Award Images (such as medal images) were reduced to 30x60 pixels. This meant that neither landscape nor square Award Images were supported. mhq now supports portrait, landscape, and square Award Images for Medals and Misc Awards (the type of award, not the autoAward type described above).

Furthermore, I have added documentation for multiple Award Images to be assigned to the same Award. This has already been described under the `Campaign Options` section.

## Closes
closes #851
closes #871
closes #2429 (or, at least, makes it no longer relevant)
closes #2433 
closes #2772
closes #2773
closes #3495
closes #3623
closes #4024

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/TheatreOfWarAwards.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+     * @param contractStartYear the contract's start yet
+     * @param currentYear the current campaign year
+     */
+    private boolean isDuringWartime (List<String> wartime, int contractStartYear, int currentYear) {
+        int contractLength = currentYear - contractStartYear;
+
+        try {
+            return IntStream.rangeClosed(0, contractLength).map(year -> contractStartYear + year)
+                    .anyMatch(checkYear -> (checkYear >= Integer.parseInt(wartime.get(0)))
+                            && (checkYear <= Integer.parseInt(wartime.get(1))));
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/TheatreOfWarAwards.java:122]: Already wrapped in try-catch

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/TheatreOfWarAwards.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+     * @param wartime a list with two entries, war start year and war end year (can be identical)
+     * @param contractStartYear the contract's start yet
+     * @param currentYear the current campaign year
+     */
+    private boolean isDuringWartime (List<String> wartime, int contractStartYear, int currentYear) {
+        int contractLength = currentYear - contractStartYear;
+
+        try {
+            return IntStream.rangeClosed(0, contractLength).map(year -> contractStartYear + year)
+                    .anyMatch(checkYear -> (checkYear >= Integer.parseInt(wartime.get(0)))
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/TheatreOfWarAwards.java:121]: Already wrapped in try-catch

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/ScenarioKillAwards.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.ResourceBundle;
+
+public class ScenarioKillAwards {
+    /**
+     * This function loops through Kill Awards, checking whether the person is eligible to receive each type of award
+     * @param campaign the campaign to be processed
+     * @param awards the awards to be processed (should only include awards where item == Kill)
+     * @param person the person to check award eligibility for
+     */
+    public ScenarioKillAwards(Campaign campaign, List<Award> awards, Person person) {
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/ScenarioKillAwards.java]: Outdated (commenting so the Review can be closed)

[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/ScenarioKillAwards.java]: Outdated (commenting so the Review can be closed)

[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/ScenarioKillAwards.java]: Outdated (commenting so the Review can be closed)

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/TrainingAwards.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.ResourceBundle;
+
+public class TrainingAwards {
+    /**
+     * This function loops through Training Awards, checking whether the person is eligible to receive each type of award
+     * @param campaign the campaign to be processed
+     * @param awards the awards to be processed (should only include awards where item == Kill)
+     * @param person the person to check award eligibility for
+     */
+    public TrainingAwards(Campaign campaign, List<Award> awards, Person person) {
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/TrainingAwards.java]: Outdated (commenting so the Review can be closed)

[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/TrainingAwards.java]: Outdated (commenting so the Review can be closed)

[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/TrainingAwards.java]: Outdated (commenting so the Review can be closed)

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/TrainingAwards.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        final ResourceBundle resource = ResourceBundle.getBundle("mekhq.resources.AutoAwards",
+                MekHQ.getMHQOptions().getLocale());
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/TrainingAwards.java]: Outdated (commenting so the Review can be closed)

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/InjuryAwards.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.List;
+
+public class InjuryAwards {
+    /**
+     * This function loops through Injury Awards, checking whether the person is eligible to receive each type of award
+     * @param campaign the campaign to be processed
+     * @param awards the awards to be processed (should only include awards where item == Kill)
+     * @param person the person to check award eligibility for
+     */
+    public InjuryAwards(Campaign campaign, List<Award> awards, Person person) {
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/InjuryAwards.java]: Resolved

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/AutoAwardsPostScenarioController.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    ArrayList<Award> injuryAwards = new ArrayList<>();
+
+    final ResourceBundle resource = ResourceBundle.getBundle("mekhq.resources.AutoAwards",
+            MekHQ.getMHQOptions().getLocale());
+
+
+    /**
+     * This creates a list of Kill Awards where item == scenario. It then loops through these Awards checking eligibility
+     *
+     * @param campaign    the campaign to be processed
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/AutoAwardsPostScenarioController.java]: Outdated (commenting so review can be c closed)

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/AutoAwardsPostScenarioController.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+     * @param scenarioId  the Id number for the scenario just completed
+     * @param person      the person to check award eligibility for
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/AutoAwardsPostScenarioController.java]: Outdated (commenting so review can be c closed)

**Comment by rroby1** [comment]
I have been waiting for this functionality ever since awards were introduced. This will be very helpful for those of us who use them. Are there plans to add an option to turn this functionality on / off for those who don't want to use it? I can imagine there would be some who don't.

**Comment by IllianiCBT** [comment]
> I have been waiting for this functionality ever since awards were introduced. This will be very helpful for those of us who use them. Are there plans to add an option to turn this functionality on / off for those who don't want to use it? I can imagine there would be some who don't. 

Absolutely, so this PR (as it stands at time of writing) is Phase One of implementation. Phase Two is all about going in and adding relevant Campaign Options options so people can tweak their experience.

File: MekHQ/src/mekhq/campaign/CampaignOptions.java [hunk 9]
```diff
@@ -2205,6 +2237,206 @@
...
+    /**
+     * @return whether to issue only the best award, if personnel are eligible for multiple awards of
+     * the same type
+     */
+    public boolean isIssueBestAwardOnly() {
+        return issueBestAwardOnly;
+    }
+
+    /**
+     * @param enable whether to issue only the best award or not
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/CampaignOptions.java:2265]: Resolved

File: MekHQ/src/mekhq/campaign/CampaignOptions.java [hunk 10]
```diff
@@ -2205,6 +2237,206 @@
...
+
+    /**
+     * @return whether to track award eligibility for this award type
+     */
+    public boolean isEnableInjuryAwards() {
+        return enableInjuryAwards;
+    }
+
+    /**
+     * @param enableFactionHunterAwards true to enable tracking of award eligibility for awards of this type
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/CampaignOptions.java:2336]: Resolved

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/AutoAwardsPostScenarioController.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        }
+
+        processBestAward(eligibleAwards);
+    }
+
+    /**
+     * If the relevant Campaign Option is enabled, this function will find the best Award Person is eligible for.
+     * It will then post eligibility to the Daily Report pane
+     * Otherwise, it will just post eligibility to the Daily Report pane for all eligible Awards
+     * @param award the award to be announced
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/AutoAwardsPostScenarioController.java:159]: Resolved

File: MekHQ/src/mekhq/gui/dialog/AutoAwardsDialog.java [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+        cboPersonnelFilter.addActionListener(evt -> filterPersonnel(personnelSorter, cboPersonnelFilter));
+
+        personnelTable.getSelectionModel().addListSelectionListener(ev -> {
+            if (personnelTable.getSelectedRow() < 0) {
+                return;
+            }
+            // TODO do I need this?
+            int rowIndex = personnelTable.convertRowIndexToModel(personnelTable.getSelectedRow());
+            UUID id = ((AutoAwardsTableModel)(personnelTable.getModel())).getPerson(rowIndex).getId();
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/gui/dialog/AutoAwardsDialog.java]: Resolved

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/KillAwards.java [hunk 13]
```diff
@@ -0,0 +1,416 @@
...
+import mekhq.campaign.personnel.Person;
+import org.apache.logging.log4j.LogManager;
+
+import java.text.MessageFormat;
+import java.util.*;
+
+public class KillAwards {
+    /**
+     * This function loops through Kill Awards, checking whether the person is eligible to receive each type of award
+     * @param c the campaign to be processed
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/KillAwards.java]: Resolved

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/KillAwards.java [hunk 14]
```diff
@@ -0,0 +1,416 @@
...
+     * @param m the mission just completed
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/KillAwards.java]: Resolved

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/KillAwards.java [hunk 15]
```diff
@@ -0,0 +1,416 @@
...
+     * @param awards the awards to be processed (should only include awards where item == Kill)
+     * @param p the person to check award eligibility for
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/KillAwards.java]: Resolved

**Comment by rroby1** [comment]
Will this work with user award sets?  The xml file will of course need to match the new format you've been working on.

**Comment by IllianiCBT** [comment]
> Will this work with user award sets? The xml file will of course need to match the new format you've been working on.

Yup! autoAwards has been written with Custom Award Sets in mind. In fact, a number of Award Types that autoAwards knows how to handle aren’t actually used by the Standard Set.

File: MekHQ/src/mekhq/campaign/CampaignOptions.java [hunk 16]
```diff
@@ -2205,6 +2243,249 @@ public void setAgeRangeRandomDeathFemaleValues(final Map<TenYearAgeRange, Double
         this.ageRangeRandomDeathFemaleValues = ageRangeRandomDeathFemaleValues;
     }
     //endregion Death
+
+    //region Awards
+    /**
+     * @return whether to issue awards to dead personnel
+     */
+    public boolean isIssuePosthumousAwards() {
```
Comment (Sleet01) [inline MekHQ/src/mekhq/campaign/CampaignOptions.java:2251]: I believe the code style recommends using 'getX' and 'setX' for getters and setters of private members; even if not, 'getX'/'setX' is the standard nomenclature.

File: MekHQ/src/mekhq/campaign/CampaignOptions.java [hunk 17]
```diff
@@ -2205,6 +2243,249 @@ public void setAgeRangeRandomDeathFemaleValues(final Map<TenYearAgeRange, Double
...
+
+    /**
+     * @param issueBestAwardOnly whether to issue only the best award or not
+     */
+    public void setIssueBestAwardOnly(final boolean issueBestAwardOnly) {
+        this.issueBestAwardOnly = issueBestAwardOnly;
+    }
+
+    /**
+     * @return whether to issue only the best award, if personnel are eligible for multiple awards of
```
Comment (Sleet01) [inline MekHQ/src/mekhq/campaign/CampaignOptions.java:2278]: Comment copy-paste

File: MekHQ/src/mekhq/campaign/CampaignOptions.java [hunk 18]
```diff
@@ -2205,6 +2243,249 @@ public void setAgeRangeRandomDeathFemaleValues(final Map<TenYearAgeRange, Double
...
+     * the same type
+     */
+    public boolean isIgnoreStandardSet() {
+        return ignoreStandardSet;
+    }
+
+    /**
+     * @param ignoreStandardSet whether to issue only the best award or not
```
Comment (Sleet01) [inline MekHQ/src/mekhq/campaign/CampaignOptions.java:2286]: Comment copy-paste

**Comment by IllianiCBT** [comment]
autoAwards is now in its' final form (barring edits required during the review phase).

All requested additions have been made, except Unit Tests. I gave them an earnest effort, but right now writing Unit Tests is beyond me.

File: MekHQ/src/mekhq/campaign/personnel/autoAwards/MiscAwards.java [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+     * @param award the award being considered
+     * @param person the unique identifier of the person
+     * @param killCount the number of kills achieved by the person (nullable)
+     * @param injuryCount the number of injuries suffered by the person (nullable)
+     * @return true if the person is eligible for the Medal of Honor award, false otherwise
+     */
+    private static boolean MedalOfHonor(Campaign campaign, Award award, UUID person, @Nullable Integer killCount, @Nullable Integer injuryCount) {
+        if (award.canBeAwarded(campaign.getPerson(person))) {
+            if ((killCount != null) && (injuryCount != null)) {
+                return (killCount >= Integer.parseInt(award.getSize())) && (injuryCount >= Integer.parseInt(award.getRange()));
```
[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/MiscAwards.java]: Resolved

[the remark this answers is not in the document]
Comment (IllianiCBT) [reply MekHQ/src/mekhq/campaign/personnel/autoAwards/MiscAwards.java]: Resolved

File: MekHQ/src/mekhq/campaign/CampaignOptions.java (hunk 16 above)
Comment (Sleet01) [reply MekHQ/src/mekhq/campaign/CampaignOptions.java:2228]: The style guide refers to "getters and setters", so I would go with 'getX' / 'setX', although apparently a '90s-era JavaBeans style guide also allows isX for getting `boolean` values.

I guess if we have a mix of getX / isX in other parts of the code, there's no harm having it here.

**Review by HammerGS — APPROVED** [review]
I've been testing this will saved games from Issues and so far haven't had an crashes. But not sure about logic bugs. Given the size and complexity looking to merge based on that and the various code QLs

**Comment by BullseyeSmith** [comment]
Daniel, you are my hero! Thank you for building a reality on the old meager foundations I left behind of my dream. I love the innovative ways you figured out how to make this work as I read all your answers to problems I couldn't solve. I don't know if the community fully realizes how powerful this will be for building investment in the growth of our personnel roster, but I do. I haven't been this excited in a very long time!

**Pull request merged**